Reactor unobserve

colindresj opened the pull request:
Creates reactor#unobserve and KeyPath.eql, which tests if two keypaths are the same.

jordangarcia commented:
I will check this out tomorrow and play around.  Given tests are passing it LGTM but I'd like to mess around with it first hand before merging.

colindresj commented:
Cool, not sure on the `KeyPath.eql` name tbh. Maybe `Keypath.same`, or something like that.

jordangarcia commented on src/key-path.js:
```diff
@@ -6,9 +6,36 @@ var isFunction = require('./utils').isFunction
+  }
+
+  while (i--) {
+    if (keyPath[i] !== otherKeyPath[i]) {
```
This needs to be recursive since getters can be composed of other getters.

It may be good to rely on the getetr hashCode, although I am a bit hesitant as others have pointed out that there is a possibility of collisions.

colindresj replied:
I might be missing something, but this is just for comparing array keypaths of `Array [ String, ...]`. It uses `isKeyPath` to make sure of that.

The actual getter comparison is an `===` reference check, since we have access to the observer entry: https://github.com/optimizely/nuclear-js/pull/115/files#diff-889e806243cf10b2e10689e9632f17a2R132.

[Here](https://github.com/optimizely/nuclear-js/pull/115/files#diff-7e02ba8688bee6af79401977651e94b9R284)'s the test for getters, using `totalGetter`, which is composed of other getters.

jordangarcia commented on src/reactor.js:
```diff
@@ -92,6 +92,48 @@ class Reactor {
+    }
+
+    each(this.__changeObserver.__observers.map(entry => {
+      if (keyPath && KeyPath.eql(keyPath, Getter.getDeps(entry.getter)[0])) {
```
Kind of confused what this line is doing.

If the getter is a keyPath then we are comparing against the first dep of the `enter.getter`?

It seems like this should be something like:

```js
const isKeyPath = keyPath.isKeyPath()

// unwatchifShould invokes entry.unwatchFn if found

each(this.__changeObserver.__observers.map(entry => {
  if (isKeyPath && KeyPath.isKeyPath(entry.getter) && KeyPath.eql(getter, entry.getter)) {
    return unwatchIfShould(entry)
  } 
  if (entry.getter === getter) {
    return unwatchIfShould(entry)
  }
}), unwatchFn => unwatchFn && unwatchFn())
```

Also are you using `map` and `each` as to not mutate the `__observers` map while iterating?

Lastly I think this logic should be in the `ChangeObserver` under `__changeObserver.unobserve`.  IMO the reactor shouldn't have knowledge of the inner workings of `ChangeObserver`

colindresj replied:
Good point. I'll update and move logic to the change observer.

bhamodi commented:
Hey @colindresj, what's the status of this PR?

colindresj commented:
@bhamodi sorry slipped my mind I had this PR open. I'll work on the stuff I said I would shortly, and will push. Thanks for the reminder

colindresj commented on src/change-observer.js:
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
+   * @param {KeyPath|Getter} getter
+   * @param {function} handler
+   */
+  unwatch(getter, handler) {
```
Moved all the logic into the changeObserver

colindresj commented on src/change-observer.js:
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
+    // Collects all the unwatchFns that need to be called, without invoking
+    // them, so as to not mutate the observers collection, then invokes each.
+    each(this.__observers.map(entry => {
+      if (isKeyPath && Getter.wasKeyPath(entry.getter)) {
```
Checking on whether both the passed getter is a keypath and if the entry's registered getter was originally a keypath, in which case we can access the original keypath using `entry.getter[0]`, since keypaths are transformed to getters using `[keypath, identity]`

colindresj commented on src/change-observer.js:
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
+        }
+      }
+
+      if (entry.getter === getter) {
```
If passed getter is a getter, we can do a simple referential equality check

colindresj commented on src/change-observer.js:
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
+    // them, so as to not mutate the observers collection, then invokes each.
+    each(this.__observers.map(entry => {
+      if (isKeyPath && Getter.wasKeyPath(entry.getter)) {
+        if (KeyPath.same(getter, entry.getter[0]) || KeyPath.isUpstream(getter, entry.getter[0])) {
```
Will unwatch when the passed keypath is the same as the entry's, or if it is upstream the entry's. IE: `['key']` is upstream `['key', 'path']`.

colindresj commented:
@bhamodi @jordangarcia Long time coming, but updated this PR. Added even more tests to cover some potential edge cases, and updated logic based on suggestions as much I could.

bhamodi commented:
@colindresj - Sorry for the delay in getting back to you. This PR now requires a rebase.
@jordangarcia - Could you review this PR one more time before we merge it in?

colindresj commented:
Rebased @bhamodi

bhamodi commented:
Thanks for doing this @colindresj! This LGTM.
Waiting to hear from @jordangarcia and then we'll merge it in.

bhamodi commented:
This has been implemented in v1.2.0 by @jordangarcia. Thanks @colindresj.
